### DISCUSSION TITLE
How to display page title in layout with app router?

### DISCUSSION THREAD

**Discussion by andreynovikov** [vercel/next.js#59532] [Help] [answered]
### Summary

I've searched the docs, asked the Goggle, have found many similar questions, but no appropriate answer. It looks like **I'm missing some major concept** of the new app router. Currently I have one layout and lot's of pages. Layout contains header, footer, sidebar, etc. Somewhere deep in the header there is a `<h1>` containing a page title. With pages router setting this title is extremely simple:

```javascript
SomePage.getLayout = function getLayout(page) {
    return (
        <MainLayout title="Page title">
            {page}
        </MainLayout>
    )
}
```

as well as for dynamic routes:

```javascript
DynamicPage.getLayout = function getLayout(page) {
    return (
        <MainLayout title={page.props.title}>
            {page}
        </MainLayout>
    )
}
```

But I couldn't find any clue how to implement it with the app router.

**Comment by pybern** [comment]
Layout is just a common UI that is shared across pages. Also `layout` only has access to `params`, if your page title can be mapped from there, it is something you can use as part of the layout. Else you might want to render your page title in `pages`.

Read more; https://nextjs.org/docs/app/api-reference/file-conventions/layout

**Comment by devbymak** [comment]
after using metadata to set the title for each page, you can access your page title in client side layout or other components with `document.title`,

for passing undefined document error we should use it after rendering the dom with the below trick:
```
  const [isMounted, setIsMounted] = useState(false)
  useEffect(() => {
    setIsMounted(true)
  }, [])

  return isMounted ? document.title : ''
```

**Reply by ivan-kleshnin** [reply]
Isn't `document.title` available only for `use client` components?

**Reply by devbymak** [reply]
yes `document` is only accessible on the client side, if we want to handle on the server side we need to create a layout for each page.

**Reply by ivan-kleshnin** [reply]
Yeah... No convenient option, unfortunately. Some people also use parallel routes for this task, but it's super verbose to create a couple of extra folders and files for a single dynamic title.

**Reply by Esensats** [reply]
Also using parallel routes for this simple task will lead to not being able to navigate away from a `not-found` page with a lot of undocumented runtime errors. Next.js, very production ready framework 👍👍

**Comment by andreynovikov** [comment, marked as answer]
@devbymak [solution](https://github.com/vercel/next.js/discussions/59532#discussioncomment-8316552) works perfectly when metadata is [statically defined](https://nextjs.org/docs/app/api-reference/functions/generate-metadata#the-metadata-object). But when [`generateMetadata()`](https://nextjs.org/docs/app/api-reference/functions/generate-metadata#generatemetadata-function) is used, title is not set. This is because first render is done before dynamic metadata is applied. So I had to implement MutationObserver hook:

```typescript
"use client";

import { useEffect, useMemo, useState } from "react";

// This is optional and can be omitted
export type TitleObserverCallback = (title: string) => void;

export function useTitleObserver(onTitleChange?: TitleObserverCallback) {
  const [title, setTitle] = useState("");

  const isMounted = useMemo(() => true, []);

  useEffect(() => {
    const observer = new MutationObserver((mutations) => {
      let newTitle = undefined;
      mutations.forEach((mutation) => {
        if (mutation.target === document.head && mutation.type === "childList") {
          mutation.addedNodes.forEach((node) => {
            if (node.nodeName === "TITLE")
              newTitle = node.textContent;
          })
        }
        if (mutation.target.nodeName === "TITLE")
          newTitle = mutation.target.textContent;
      });
      if (newTitle !== undefined)
        setTitle(newTitle);
    });

    // <head> is observed because Next.js does not change title contents but removes and adds <title> node
    observer.observe(document.head, {
      subtree: true,
      characterData: true,
      childList: true,
    });

    return () => observer.disconnect();
  }, []);

  // As mentioned earlier this is optional and can be omitted
  useEffect(() => {
    if (title !== "" && onTitleChange)
      onTitleChange(title);
  }, [onTitleChange, title]);

  // Return dynamically set title
  if (title !== "")
    return title;

  // Fallback to static title
  return isMounted ? document.title : "";
}
```

**Reply by devbymak** [reply]
Solution works, but it’s not recommended. It relies on internal Next.js `<head>` behavior, is fragile across updates, and adds unnecessary complexity. In the App Router, the correct approach is to treat the title as shared data (derive it once and use it for both generateMetadata and your layout).
